src/Nethermind/Nethermind.TxPool/TxNonceTxPoolReserveSealer.cs

```diff
@@ -18,25 +18,36 @@
 using System;
 using System.Threading.Tasks;
 using Nethermind.Core;
+using Nethermind.Core.Specs;
+using Nethermind.Crypto;
+using Nethermind.Logging;
 
 namespace Nethermind.TxPool
 {
     // TODO: this should be nonce reserving tx sender, not sealer?
     public class NonceReservingTxSealer : TxSealer
     {
         private readonly ITxPool _txPool;
+        private readonly IEthereumEcdsa _ecdsa;
 
-        public NonceReservingTxSealer(ITxSigner txSigner, ITimestamper timestamper, ITxPool txPool)
+        public NonceReservingTxSealer(ITxSigner txSigner,
+            ITimestamper timestamper,
+            ITxPool txPool,
+            IEthereumEcdsa ecdsa)
             : base(txSigner, timestamper)
         {
             _txPool = txPool ?? throw new ArgumentNullException(nameof(txPool));
+            _ecdsa = ecdsa ?? throw new ArgumentNullException(nameof(ecdsa));
         }
 
         public override ValueTask Seal(Transaction tx, TxHandlingOptions txHandlingOptions)
         {
             bool manageNonce = (txHandlingOptions & TxHandlingOptions.ManagedNonce) == TxHandlingOptions.ManagedNonce;
             if (manageNonce)
             {
+                tx.SenderAddress ??= _ecdsa.RecoverAddress(tx);
```

Comment:
@smartprogrammer93 I'm still don't get it. Before your refactoring we didn't need this recovery and we were passing all hive tests. Later, we started failing some hive tests, so we decided to add RecoverAddress here, but why? what changes caused the issue? 

Not saying that it is bad, but I want to understand the changes and be sure that you understand it too.